src/main/java/org/truffleruby/core/constant/WarnAlreadyInitializedNode.java

```diff
@@ -21,13 +21,23 @@
     @Child private WarnNode warnNode = new WarnNode();
 
     @TruffleBoundary
-    public void warnAlreadyInitialized(DynamicObject module, String name, SourceSection sourceSection) {
+    public void warnAlreadyInitialized(DynamicObject module, String name, SourceSection sourceSection, SourceSection previousSourceSection) {
         final String moduleName = Layouts.MODULE.getFields(module).getName();
+        final String constName;
+        if (module == getContext().getCoreLibrary().getObjectClass()) {
+            constName = name;
+        } else {
+            constName = moduleName + "::" + name;
+        }
         if (sourceSection != null) {
             warnNode.warn(sourceSection.getSource().getName(), ":", Integer.toString(sourceSection.getStartLine()),
-                    ": warning: already initialized constant ", moduleName, "::", name);
+                    ": warning: already initialized constant ", constName);
         } else {
-            warnNode.warn("warning: already initialized constant ", moduleName, "::", name);
+            warnNode.warn("warning: already initialized constant ", constName);
```

Comment:
To avoid a bit duplication here, could we just have one call to warnNode.warn() and extract the source information in a String variable, by default empty? (String concatenation is OK here I think for formatting a SourceSection)